lxd/storage_zfs_utils.go

```diff
@@ -89,6 +89,23 @@ func zfsPoolCheck(pool string) error {
 	return nil
 }
 
+// zfsPoolVolumeExists verifies if a specific ZFS pool or volume exists.
+func zfsPoolVolumeExists(dataset string) (bool, error) {
+	output, err := shared.RunCommand(
+		"zfs", "list", "-Ho", "name")
```

Comment:
I'm not familiar with this code base ... I just peeked into this PR from the weekly news update out of curiosity ... so I might be missing something. :sweat_smile: 

Is there a specific reason why you don't also pass the `dataset` to `zfs list`? `zfs list` exits with a non-zero code when the dataset doesn't exists. You could ignore the output and just decide from the return code.

Reply:
So my guess is that this was done so we don't need to deal with parsing error messages from the ZFS tools.

If calling directly `zfs get` or `zfs list` for a given dataset, the error message in case it doesn't exist may differ based on version or may even be translated. It also doesn't appear that ZFS gives you a different return code in such cases.

Calling `zfs list` for all datasets means that it will only fail if something else is wrong (something which should be reported as an error) and will otherwise succeed, potentially giving an empty list.

We can then reliably check against that list and avoid false positives/negatives.

Reply:
@dnegreira Hmm ... I think I understand. I wrongly assumed `zfs list` would only fail if the dataset didn't exist. So by doing it this way you made sure that you can cleanly separate "something is wrong with the zfs command" and "dataset doesn't exist." :+1:

Reply:
Yep, exactly!